src/python/pants/build_graph/import_remote_sources_mixin.py

```diff
@@ -0,0 +1,124 @@
+# coding=utf-8
+# Copyright 2015 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from builtins import str
+
+from future.utils import string_types
+
+from pants.build_graph.address_lookup_error import AddressLookupError
+from pants.build_graph.target import Target
+from pants.util.memo import memoized_property
+from pants.util.meta import AbstractClass, classproperty
+from pants.util.objects import TypeConstraintError
+
+
+class ImportRemoteSourcesMixin(Target, AbstractClass):
+  """A Target Mixin to be used when a target declares another target type to be imported."""
+
+  class ExpectedAddressError(AddressLookupError):
+    """Thrown upon attempting to resolved an object that is not an address."""
+
+  class UnresolvedImportError(AddressLookupError):
+    """Raised when an imported target cannot be resolved."""
+
+  class WrongTargetTypeError(AddressLookupError):
+    """Thrown if the wrong type of target is encountered."""
+
+  @classproperty
+  def expected_target_constraint(cls):
+    """
+    :returns: A type constraint which is used to validate the targets containing remote sources,
+              specified `imported_target_kwargs_field` in a BUILD file.
+    :rtype: TypeConstraint
+    """
+    # TODO: make a utility in pants.util.meta to register these standard `NotImplementedError`s!
+    raise NotImplementedError(
+      'subclasses of ImportRemoteSourcesMixin must implement an '
+      '`expected_target_constraint` classproperty'
+    )
+
+  @classproperty
+  def imported_target_kwargs_field(cls):
```

Comment:
It feels like it should be possible to standardize this rather than making it optional...? (especially without java_protobuf_library in the mix).

Reply:
Turned into #7111!